middleware/client_ip.go

```diff
@@ -0,0 +1,185 @@
+		}
+		return http.HandlerFunc(fn)
+	}
+}
+
+// ClientIPFromXFFHeader parses the client IP address from the X-Forwarded-For
+// header and injects it into the request context if it is not already set. The
+// parsed IP address can be retrieved using GetClientIP().
+//
+// The middleware traverses the X-Forwarded-For chain (rightmost untrusted IP)
+// and excludes loopback, private, unspecified, and trusted IP ranges.
+//
+// ### Important Notice:
+// - Use this middleware only when your infrastructure sets and validates the X-Forwarded-For header.
+// - Malicious clients can spoof the header unless a trusted reverse proxy or load balancer sanitizes it.
+//
+// Parameters:
+// - `trustedIPPrefixes`: A list of CIDR prefixes that define trusted proxy IP ranges.
+//
+// Example trusted IP ranges:
+// - "203.0.113.0/24"     - Example corporate proxy
+// - "198.51.100.0/24"    - Example data center or hosting provider
+// - "2400:cb00::/32"     - Cloudflare IPv6 range
+// - "2606:4700::/32"     - Cloudflare IPv6 range
+// - "192.0.2.0/24"       - Example VPN gateway
+//
+// Note: Private IP ranges (e.g., "10.0.0.0/8", "192.168.0.0/16", "172.16.0.0/12")
+// are automatically excluded by netip.Addr.IsPrivate() and do not need to be added here.
+func ClientIPFromXFFHeader(trustedIPPrefixes ...string) func(http.Handler) http.Handler {
+	// Pre-parse trusted prefixes.
+	trustedPrefixes := make([]netip.Prefix, len(trustedIPPrefixes))
+	for i, ipRange := range trustedIPPrefixes {
+		trustedPrefixes[i] = netip.MustParsePrefix(ipRange)
+	}
+
+	return func(h http.Handler) http.Handler {
+		fn := func(w http.ResponseWriter, r *http.Request) {
+			ctx := r.Context()
+
+			// Check if the client IP is already set in the context.
+			if _, ok := ctx.Value(clientIPCtxKey).(netip.Addr); ok {
+				h.ServeHTTP(w, r)
+				return
+			}
+
+			// Parse and split the X-Forwarded-For header(s).
+			xff := strings.Split(strings.Join(r.Header.Values("X-Forwarded-For"), ","), ",")
+		nextValue:
+			for i := len(xff) - 1; i >= 0; i-- {
+				ip, err := netip.ParseAddr(strings.TrimSpace(xff[i]))
+				if err != nil {
+					continue
+				}
+
+				// Ignore loopback, private, or unspecified addresses.
+				if ip.IsLoopback() || ip.IsPrivate() || ip.IsUnspecified() {
+					continue
+				}
+
+				// Ignore trusted IPs within the given ranges.
+				for _, prefix := range trustedPrefixes {
+					if prefix.Contains(ip) {
+						continue nextValue
+					}
+				}
+
+				// Store the valid client IP in the context.
+				ctx = context.WithValue(ctx, clientIPCtxKey, ip)
+				h.ServeHTTP(w, r.WithContext(ctx))
+				return
+			}
+
+			h.ServeHTTP(w, r)
+		}
+		return http.HandlerFunc(fn)
+	}
+}
+
+// ClientIPFromRemoteAddr extracts the client IP address from the RemoteAddr
+// field of the HTTP request and injects it into the request context if it is
+// not already set. The parsed IP address can be retrieved using GetClientIP().
+//
+// The middleware ignores invalid or private IPs.
```

Comment:
As above: Why? The user has indicated that this is the value they want.

But also... I don't think you're actually doing the loopback/private/unspecified check here, so I don't think this comment is true?